### PR TITLE
Add BindAddress and PublicAddress Config options

### PR DESCRIPTION
**Pull request by craftxbox**
### Purpose
This PR allows server hosts to specify which Interface Address to bind to, instead of the default of *all* addresses.
Aswell, This allows a server host to change the IP address their server presents to the Hub, if the IP address lookup result gets the wrong address (IE, on multi-address systems).
This also allows server hosts to put a *domain* instead of an IP address in their hub listing, Potentially allowing extra smartness in the launcher with DNS SRV, A, and AAAA records.

### Notes:

**IMPORTANT**: Domains in hub listing is not supported by the hub yet (see unitystation/stationhub#161)

### Changelog:
CL: [New] Bind Address field for Server Hosts
CL: [New] Public Address field for Server Hosts

### CODE REVIEW

**Comment by craftxbox** [comment]
State: Tested functional, IPV6 support hasn't worked but i'm certain it's a networking issue on my side.
Netstat shows it binds to V6 addresses as expected, and local-loopback connections have worked fine, Just not remote ones.

File: UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs [hunk 1]
```diff
@@ -96,9 +96,20 @@ IEnumerator SendServerStatus()
...
+					{
+						publicIP = config.PublicAddress;
+					} 
+					else if (string.IsNullOrEmpty(config.BindAddress) == false)
+					{
+						publicIP = config.BindAddress;
+					}
+					else
+					{
+						req = UnityWebRequest.Get("http://ipinfo.io/ip");
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs:109]: I gave a quick look at that website and it looks like a paid service?

File: UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs [hunk 2]
```diff
@@ -96,9 +96,20 @@ IEnumerator SendServerStatus()
...
+						yield return req.SendWebRequest();
+						publicIP = Regex.Replace(req.downloadHandler.text, @"\t|\n|\r", "");
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs:113]: by convention, always add a comment explaining the regex in plain English

File: UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs [hunk 3]
```diff
@@ -164,13 +164,23 @@ void ApplyConfig()
...
 			if (ignorance != null)
 			{
-				ignorance.port = (ushort) config.ServerPort;
+				ignorance.port = (ushort)config.ServerPort;
+
+			}
+		}
+		if (string.IsNullOrEmpty(config.BindAddress) == false)
+		{
+			var ignorance = GetComponent<Ignorance>();
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs:179]: Must be a generic way of setting this?

File: UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs (hunk 2 above)
Comment (craftxbox) [reply UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs:113]: Not my code

File: UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs (hunk 1 above)
Comment (craftxbox) [reply UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs:109]: Not my code

Comment (craftxbox) [reply UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs:109]: The /ip endpoint just returns any public ip, and it's been in the code for a loong time now so I don't see any reason to change it unless there's a more guaranteed alternative.

File: UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs (hunk 2 above)
Comment (craftxbox) [reply UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs:113]: `/\t|\n|\r/` -> Match Tab, Newline, or Carriage Return.

File: UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs (hunk 3 above)
Comment (craftxbox) [reply UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs:179]: I can move the variable declaration outside if that's what you mean?

File: UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs (hunk 2 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Core/Database/ServerData.ServerStatus.cs:113]: I understand you just moved it, but can you improve it and add the comment, pretty please?

File: UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs (hunk 3 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs:179]: @Bod9001 do you mean Telepathy and Ignorance extend the same class?

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs:179]: Doesn't ignorance use a generic networking layer so mirror works with it, that's what I mean

Comment (craftxbox) [reply UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs:179]: No, it can't be abstracted to that. Ignorance is the transport, which is below mirror. Ignorance does the listening (with ENet), Mirror is just the protocol.

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Managers/NetworkManagement/CustomNetworkManager.cs:179]: ok cool

**Review by Bod9001 — DISMISSED** [review]

**Review by NoooneyDude — DISMISSED** [review]

**Comment by NoooneyDude** [comment]
Not interested in satisfying Codacy?

**Comment by corp-0** [comment]
before merging, please update the pr to latest develop so that the new workflow for pr2changelog runs fine

**Review by corp-0 — APPROVED** [review]

**Review by NoooneyDude — APPROVED** [review]

**Comment by craftxbox** [comment]
> Not interested in satisfying Codacy?

**Comment by craftxbox** [comment]
> Not interested in satisfying Codacy?

Not in the slightest, it's suggesting changes that go against the existing style in the file and I don't want to jumble it up.
